### PR TITLE
[SPARK-39217][SQL] Makes DPP support the pruning side has Union

### PR DESCRIPTION
**Pull request by wankunde**
### What changes were proposed in this pull request?
Makes DPP support the pruning side has `Union`. For example:

```sql
SELECT f.store_id,
       f.date_id,
       s.state_province
FROM (SELECT 4 AS store_id,
               date_id,
               product_id
      FROM   fact_sk
      WHERE  date_id >= 1300
      UNION ALL
      SELECT   store_id,
               date_id,
               product_id
      FROM   fact_stats
      WHERE  date_id <= 1000) f
JOIN dim_store s
ON f.store_id = s.store_id
WHERE s.country IN ('US', 'NL')
```

After this PR: image

### Why are the changes needed?
Improve query performance.

### Does this PR introduce _any_ user-facing change?
No.

### How was this patch tested?
Unit test.

### CODE REVIEW

**Review by wangyum — APPROVED** [review]

**Review by dongjoon-hyun — COMMENTED** [review]
Could you re-trigger failed tests, @wankunde ?

File: sql/core/src/test/scala/org/apache/spark/sql/DynamicPartitionPruningSuite.scala [hunk 1]
```diff
@@ -1616,6 +1616,36 @@ abstract class DynamicPartitionPruningSuiteBase
       assert(collectDynamicPruningExpressions(df.queryExecution.executedPlan).size === 1)
     }
   }
+
+    test("SPARK-39217: Makes DPP support the pruning side has Union") {
```
Comment (dongjoon-hyun) [inline sql/core/src/test/scala/org/apache/spark/sql/DynamicPartitionPruningSuite.scala:1620]: indentation?

**Comment by dongjoon-hyun** [comment]
cc @sunchao

File: sql/core/src/test/scala/org/apache/spark/sql/DynamicPartitionPruningSuite.scala (hunk 1 above)
Comment (wankunde) [reply sql/core/src/test/scala/org/apache/spark/sql/DynamicPartitionPruningSuite.scala:1620]: Fixed, thanks

**Comment by wangyum** [comment]
cc @cloud-fan

**Review by dongjoon-hyun — APPROVED** [review]

**Comment by wangyum** [comment]
Merged to master.

**Comment by cloud-fan** [comment]
late LGTM

**Pull request closed without merging**